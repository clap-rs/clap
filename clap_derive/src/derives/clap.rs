// Copyright 2018 Guillaume Pinot (@TeXitoi) <texitoi@texitoi.eu>,
// Kevin Knapp (@kbknapp) <kbknapp@gmail.com>, and
// Andrew Hobden (@hoverbear) <andrew@hoverbear.org>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
// This work was derived from Structopt (https://github.com/TeXitoi/structopt)
// commit#ea76fa1b1b273e65e3b0b1046643715b49bec51f which is licensed under the
// MIT/Apache 2.0 license.

use crate::{
    derives::{arg_enum, from_argmatches, into_app, subcommand},
    dummies,
};

use proc_macro2::TokenStream;
use proc_macro_error::abort_call_site;
use quote::quote;
use syn::{self, punctuated, token, Attribute, DataEnum, DeriveInput, Field, Ident};

pub fn derive_clap(input: &DeriveInput) -> TokenStream {
    use syn::Data::*;

    let ident = &input.ident;

    match input.data {
        Struct(syn::DataStruct {
            fields: syn::Fields::Named(ref fields),
            ..
        }) => {
            dummies::clap_struct(ident);
            gen_for_struct(ident, &fields.named, &input.attrs)
        }
        Struct(syn::DataStruct {
            fields: syn::Fields::Unit,
            ..
        }) => {
            dummies::clap_struct(ident);
            gen_for_struct(
                ident,
                &punctuated::Punctuated::<Field, token::Comma>::new(),
                &input.attrs,
            )
        }
        Enum(ref e) => {
            dummies::clap_enum(ident);
            gen_for_enum(ident, &input.attrs, e)
        }
        _ => abort_call_site!("`#[derive(Clap)]` only supports non-tuple structs and enums"),
    }
}

fn gen_for_struct(
    name: &Ident,
    fields: &punctuated::Punctuated<Field, token::Comma>,
    attrs: &[Attribute],
) -> TokenStream {
    let (into_app, attrs) = into_app::gen_for_struct(name, fields, attrs);
    let from_arg_matches = from_argmatches::gen_for_struct(name, fields, &attrs);

    quote! {
        impl ::clap::Clap for #name {}

        #into_app
        #from_arg_matches
    }
}

fn gen_for_enum(name: &Ident, attrs: &[Attribute], e: &DataEnum) -> TokenStream {
    let into_app = into_app::gen_for_enum(name);
    let from_arg_matches = from_argmatches::gen_for_enum(name);
    let subcommand = subcommand::gen_for_enum(name, attrs, e);

    let arg_enum = if e.variants.iter().all(|v| match v.fields {
        syn::Fields::Unit => true,
        _ => false,
    }) {
        arg_enum::gen_for_enum(name, attrs, e)
    } else {
        quote!()
    };

    quote! {
        impl ::clap::Clap for #name { }

        #into_app
        #from_arg_matches
        #subcommand
        #arg_enum
    }
}
