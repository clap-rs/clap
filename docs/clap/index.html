<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="rustdoc">
    <meta name="description" content="API documentation for the Rust `clap` crate.">
    <meta name="keywords" content="rust, rustlang, rust-lang, clap">

    <title>clap - Rust</title>

    <link rel="stylesheet" type="text/css" href="../main.css">

    
    
</head>
<body class="rustdoc">
    <!--[if lte IE 8]>
    <div class="warning">
        This old browser is unsupported and will most likely display funky
        things.
    </div>
    <![endif]-->

    

    <section class="sidebar">
        
        <p class='location'></p><script>window.sidebarCurrent = {name: 'clap', ty: 'mod', relpath: '../'};</script>
    </section>

    <nav class="sub">
        <form class="search-form js-only">
            <div class="search-container">
                <input class="search-input" name="search"
                       autocomplete="off"
                       placeholder="Click or press 'S' to search, '?' for more options..."
                       type="search">
            </div>
        </form>
    </nav>

    <section id='main' class="content mod">
<h1 class='fqn'><span class='in-band'>Crate <a class='mod' href=''>clap</a><wbr></span><span class='out-of-band'><a href='stability.html'>[stability]</a> <span id='render-detail'>
            <a id="collapse-all" href="#">[-]</a>&nbsp;<a id="expand-all" href="#">[+]</a>
        </span><a id='src-0' href='../src/clap/lib.rs.html#1-272'>[src]</a></span></h1>
<div class='docblock'>
<h1 id="clap" class='section-header'><a
                           href="#clap">clap</a></h1>
<p><img src="https://travis-ci.org/kbknapp/clap-rs.svg?branch=master" alt="Travis-CI"></p>

<p>Command Line Argument Parser written in Rust</p>

<p>A simply library for parsing command line arguments and subcommands when writing command line and console applications.</p>

<p>You can use <code>clap</code> to lay out a list of possible valid command line arguments and subcommands, then let <code>clap</code> parse the string given by the user at runtime.</p>

<p>When using <code>clap</code> you define a set of parameters and rules for your arguments and subcommands, then at runtime <code>clap</code> will determine their validity.</p>

<p><code>clap</code> also provides the traditional version and help switches &#39;for free&#39; by parsing the list of possible valid arguments lazily at runtime, and if not already defined by the developer <code>clap</code> will autogenerate all applicable &quot;help&quot; and &quot;version&quot; switches (as well as a &quot;help&quot; subcommand if other subcommands are defined as well).</p>

<p>After defining a list of possible valid arguments and subcommands, <code>clap</code> gives you a list of valid matches that the user supplied at runtime, or informs the user of their error and exits gracefully. You can use this list to determine the functioning of your program.</p>

<h2 id="quick-example" class='section-header'><a
                           href="#quick-example">Quick Example</a></h2><pre id='rust-example-rendered' class='rust '>
<span class='comment'>// (Full example with comments in examples/01_QuickExample.rs)</span>
<span class='kw'>extern</span> <span class='kw'>crate</span> <span class='ident'>clap</span>;
<span class='kw'>use</span> <span class='ident'>clap</span>::{<span class='ident'>Arg</span>, <span class='ident'>App</span>, <span class='ident'>SubCommand</span>};
 
<span class='kw'>fn</span> <span class='ident'>main</span>() {
    <span class='kw'>let</span> <span class='ident'>matches</span> <span class='op'>=</span> <span class='ident'>App</span>::<span class='ident'>new</span>(<span class='string'>&quot;MyApp&quot;</span>)
                          .<span class='ident'>version</span>(<span class='string'>&quot;1.0&quot;</span>)
                          .<span class='ident'>author</span>(<span class='string'>&quot;Kevin K. &lt;kbknapp@gmail.com&gt;&quot;</span>)
                          .<span class='ident'>about</span>(<span class='string'>&quot;Does awesome things&quot;</span>)
                          .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;CONFIG&quot;</span>)
                               .<span class='ident'>short</span>(<span class='string'>&quot;c&quot;</span>)
                               .<span class='ident'>long</span>(<span class='string'>&quot;config&quot;</span>)
                               .<span class='ident'>help</span>(<span class='string'>&quot;Sets a custom config file&quot;</span>)
                               .<span class='ident'>takes_value</span>(<span class='boolval'>true</span>))
                          .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;output&quot;</span>)
                               .<span class='ident'>help</span>(<span class='string'>&quot;Sets an optional output file&quot;</span>)
                               .<span class='ident'>index</span>(<span class='number'>1</span>))
                          .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;debug&quot;</span>)
                               .<span class='ident'>short</span>(<span class='string'>&quot;d&quot;</span>)
                               .<span class='ident'>multiple</span>(<span class='boolval'>true</span>)
                               .<span class='ident'>help</span>(<span class='string'>&quot;Turn debugging information on&quot;</span>))
                          .<span class='ident'>subcommand</span>(<span class='ident'>SubCommand</span>::<span class='ident'>new</span>(<span class='string'>&quot;test&quot;</span>)
                                      .<span class='ident'>about</span>(<span class='string'>&quot;controls testing features&quot;</span>)
                                      .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;verbose&quot;</span>)
                                          .<span class='ident'>short</span>(<span class='string'>&quot;v&quot;</span>)
                                          .<span class='ident'>help</span>(<span class='string'>&quot;print test information verbosely&quot;</span>)))
                          .<span class='ident'>get_matches</span>();
 
    <span class='kw'>if</span> <span class='kw'>let</span> <span class='prelude-val'>Some</span>(<span class='ident'>o</span>) <span class='op'>=</span> <span class='ident'>matches</span>.<span class='ident'>value_of</span>(<span class='string'>&quot;output&quot;</span>) {
        <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Value for output: {}&quot;</span>, <span class='ident'>o</span>);
    }
 
    <span class='kw'>if</span> <span class='kw'>let</span> <span class='prelude-val'>Some</span>(<span class='ident'>c</span>) <span class='op'>=</span> <span class='ident'>matches</span>.<span class='ident'>value_of</span>(<span class='string'>&quot;CONFIG&quot;</span>) {
        <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Value for config: {}&quot;</span>, <span class='ident'>c</span>);
    }
 
    <span class='kw'>match</span> <span class='ident'>matches</span>.<span class='ident'>occurrences_of</span>(<span class='string'>&quot;debug&quot;</span>) {
        <span class='number'>0</span> <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Debug mode is off&quot;</span>),
        <span class='number'>1</span> <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Debug mode is kind of on&quot;</span>),
        <span class='number'>2</span> <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Debug mode is on&quot;</span>),
        <span class='number'>3</span> <span class='op'>|</span> _ <span class='op'>=&gt;</span> <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Don&#39;t be crazy&quot;</span>),
    }
 
    <span class='kw'>if</span> <span class='kw'>let</span> <span class='prelude-val'>Some</span>(<span class='kw-2'>ref</span> <span class='ident'>matches</span>) <span class='op'>=</span> <span class='ident'>matches</span>.<span class='ident'>subcommand_matches</span>(<span class='string'>&quot;test&quot;</span>) {
        <span class='kw'>if</span> <span class='ident'>matches</span>.<span class='ident'>is_present</span>(<span class='string'>&quot;verbose&quot;</span>) {
            <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Printing verbosely...&quot;</span>);
        } <span class='kw'>else</span> {
            <span class='macro'>println</span><span class='macro'>!</span>(<span class='string'>&quot;Printing normally...&quot;</span>);
        }
    }
 
    <span class='comment'>// more porgram logic goes here...</span>
}
</pre>

<p>If you were to compile the above program and run it with the flag <code>--help</code> or <code>-h</code> (or <code>help</code> subcommand, since we defined <code>test</code> as a subcommand) the following output woud be presented</p>

<pre><code class="language-sh">$ myprog --help
MyApp 1.0
Kevin K. &lt;kbknapp@gmail.com&gt;
Does awesome things
 
USAGE:
    MyApp [FLAGS] [OPTIONS] [POSITIONAL] [SUBCOMMANDS]
 
FLAGS:
    -d               Turn debugging information on
    -h,--help        Prints this message
    -v,--version     Prints version information
  
OPTIONS:
    -c,--config=CONFIG        Sets a custom config file
 
POSITIONAL ARGUMENTS:
    output            Sets an optional output file
 
SUBCOMMANDS:
    help            Prints this message
    test            Controls testing features
</code></pre>

<h2 id="installation" class='section-header'><a
                           href="#installation">Installation</a></h2>
<p>Add <code>clap</code> as a dependecy in your <code>Cargo.toml</code> file to use from crates.io:</p>
<pre id='rust-example-rendered' class='rust '>
 [<span class='ident'>dependencies</span>]
 <span class='ident'>clap</span> <span class='op'>=</span> <span class='string'>&quot;*&quot;</span>
</pre>

<p>Or track the latest on the master branch at github:</p>
<pre id='rust-example-rendered' class='rust '>
[<span class='ident'>dependencies</span>.<span class='ident'>clap</span>]
<span class='ident'>git</span> <span class='op'>=</span> <span class='string'>&quot;https://github.com/kbknapp/clap-rs.git&quot;</span>
</pre>

<p>Then run <code>cargo build</code> or <code>cargo update</code> for your project.</p>

<h2 id="usage" class='section-header'><a
                           href="#usage">Usage</a></h2>
<p>Add <code>extern crate clap;</code> to your crate root.</p>

<h2 id="more-information" class='section-header'><a
                           href="#more-information">More Information</a></h2>
<p>You can find complete documentation on the <a href="http://kbknapp.github.io/clap-rs/docs/clap/index.html">github-pages site</a> for this project.</p>

<p>You can also find full usage examples in the examples/ directory of this repo.</p>

<h2 id="how-to-build" class='section-header'><a
                           href="#how-to-build">How to build</a></h2>
<p>Clone the repo (<code>master</code> branch if using <code>clap</code>, <code>dev</code> branch if contributing) and build with Cargo</p>
<pre id='rust-example-rendered' class='rust '>
<span class='ident'>git</span> <span class='ident'>clone</span> <span class='ident'>https</span>:<span class='comment'>//github.com/kbknapp/clap-rs</span>
<span class='ident'>cd</span> <span class='ident'>clap</span><span class='op'>-</span><span class='ident'>rs</span>
<span class='ident'>cargo</span> <span class='ident'>build</span> <span class='op'>-</span><span class='op'>-</span><span class='ident'>release</span>
</pre>

<h3 id="running-the-tests" class='section-header'><a
                           href="#running-the-tests">Running the tests</a></h3>
<p>If contributing, you can run the tests as follows (assuming you&#39;ve already cloned the repo to <code>clap-rs/</code></p>
<pre id='rust-example-rendered' class='rust '>
<span class='ident'>cd</span> <span class='ident'>clap</span><span class='op'>-</span><span class='ident'>rs</span><span class='op'>/</span><span class='ident'>claptests</span>
<span class='ident'>make</span> <span class='ident'>test</span>
</pre>

<h3 id="building-the-documentation" class='section-header'><a
                           href="#building-the-documentation">Building the documentation</a></h3>
<p>Run this instead of <code>cargo doc</code> to generate the proper module docstring:</p>
<pre id='rust-example-rendered' class='rust '>
<span class='ident'>make</span> <span class='ident'>doc</span>
</pre>

<p>Then browse to <code>clap-rs/docs/index.html</code> in your web-browser of choice</p>
</div><h2 id='structs' class='section-header'><a href="#structs">Structs</a></h2>
<table>
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.App.html'
                               title='clap::App'>App</a></td>
                        <td class='docblock short'><p>Used to create a representation of the program and all possible command line arguments
for parsing at runtime.</p>
</td>
                    </tr>
                
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.Arg.html'
                               title='clap::Arg'>Arg</a></td>
                        <td class='docblock short'><p>The abstract representation of a command line argument used by the consumer of the library.</p>
</td>
                    </tr>
                
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.ArgMatches.html'
                               title='clap::ArgMatches'>ArgMatches</a></td>
                        <td class='docblock short'><p>Used to get information about the arguments that
where supplied to the program at runtime.</p>
</td>
                    </tr>
                
                    <tr>
                        <td><a class='stability Unmarked' title='No stability level'></a><a class='struct' href='struct.SubCommand.html'
                               title='clap::SubCommand'>SubCommand</a></td>
                        <td class='docblock short'><p>The abstract representation of a command line subcommand used by the consumer of the library.</p>
</td>
                    </tr>
                </table></section>
    <section id='search' class="content hidden"></section>

    <section class="footer"></section>

    <div id="help" class="hidden">
        <div class="shortcuts">
            <h1>Keyboard shortcuts</h1>
            <dl>
                <dt>?</dt>
                <dd>Show this help dialog</dd>
                <dt>S</dt>
                <dd>Focus the search field</dd>
                <dt>&larrb;</dt>
                <dd>Move up in search results</dd>
                <dt>&rarrb;</dt>
                <dd>Move down in search results</dd>
                <dt>&#9166;</dt>
                <dd>Go to active search result</dd>
            </dl>
        </div>
        <div class="infos">
            <h1>Search tricks</h1>
            <p>
                Prefix searches with a type followed by a colon (e.g.
                <code>fn:</code>) to restrict the search to a given type.
            </p>
            <p>
                Accepted types are: <code>fn</code>, <code>mod</code>,
                <code>struct</code>, <code>enum</code>,
                <code>trait</code>, <code>typedef</code> (or
                <code>tdef</code>).
            </p>
        </div>
    </div>

    

    <script>
        window.rootPath = "../";
        window.currentCrate = "clap";
        window.playgroundUrl = "";
    </script>
    <script src="../jquery.js"></script>
    <script src="../main.js"></script>
    
    <script async src="../search-index.js"></script>
</body>
</html>