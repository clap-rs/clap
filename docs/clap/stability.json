{"name":"clap","counts":{"deprecated":0,"unstable":0,"stable":0,"unmarked":40},"submodules":[]}