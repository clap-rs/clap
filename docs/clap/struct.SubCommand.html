<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="rustdoc">
    <meta name="description" content="API documentation for the Rust `SubCommand` struct in crate `clap`.">
    <meta name="keywords" content="rust, rustlang, rust-lang, SubCommand">

    <title>clap::SubCommand - Rust</title>

    <link rel="stylesheet" type="text/css" href="../main.css">

    
    
</head>
<body class="rustdoc">
    <!--[if lte IE 8]>
    <div class="warning">
        This old browser is unsupported and will most likely display funky
        things.
    </div>
    <![endif]-->

    

    <section class="sidebar">
        
        <p class='location'><a href='index.html'>clap</a></p><script>window.sidebarCurrent = {name: 'SubCommand', ty: 'struct', relpath: ''};</script><script defer src="sidebar-items.js"></script>
    </section>

    <nav class="sub">
        <form class="search-form js-only">
            <div class="search-container">
                <input class="search-input" name="search"
                       autocomplete="off"
                       placeholder="Click or press 'S' to search, '?' for more options..."
                       type="search">
            </div>
        </form>
    </nav>

    <section id='main' class="content struct">
<h1 class='fqn'><span class='in-band'>Struct <a href='index.html'>clap</a>::<wbr><a class='struct' href=''>SubCommand</a><wbr></span><span class='out-of-band'><span id='render-detail'>
            <a id="collapse-all" href="#">[-]</a>&nbsp;<a id="expand-all" href="#">[+]</a>
        </span><a id='src-8320' href='../src/clap/args/subcommand.rs.html#23-26'>[src]</a></span></h1>
<pre class='rust struct'>pub struct SubCommand&lt;'a&gt; {
    pub name: <a class='struct' href='http://doc.rust-lang.org/nightly/collections/string/struct.String.html' title='collections::string::String'>String</a>,
    pub matches: <a class='struct' href='../clap/struct.ArgMatches.html' title='clap::ArgMatches'>ArgMatches</a>&lt;'a&gt;,
}</pre><div class='docblock'><p>The abstract representation of a command line subcommand used by the consumer of the library.</p>

<p>This struct is used by the library consumer and describes all the valid options of the subcommand for 
their program. SubCommands are treated like &quot;sub apps&quot; and contain all the same possibilities (such as
their own arguments and subcommands).</p>

<h1 id="example" class='section-header'><a
                           href="#example">Example</a></h1><pre id='rust-example-rendered' class='rust '>
<span class='ident'>SubCommand</span>::<span class='ident'>new</span>(<span class='string'>&quot;conifg&quot;</span>)
               .<span class='ident'>about</span>(<span class='string'>&quot;Used for configuration&quot;</span>)
               .<span class='ident'>arg</span>(<span class='ident'>Arg</span>::<span class='ident'>new</span>(<span class='string'>&quot;config_file&quot;</span>)
                          .<span class='ident'>help</span>(<span class='string'>&quot;The configuration file to use&quot;</span>)
                          .<span class='ident'>index</span>(<span class='number'>1</span>))
</pre>
</div><h2 class='fields'>Fields</h2>
<table><tr><td id='structfield.name'><a class='stability Unmarked' title='No stability level'></a><code>name</code></td><td></td></tr><tr><td id='structfield.matches'><a class='stability Unmarked' title='No stability level'></a><code>matches</code></td><td></td></tr></table><h2 id='methods'>Methods</h2><h3 class='impl'><a class='stability Unmarked' title='No stability level'></a><code>impl&lt;'a&gt; <a class='struct' href='../clap/struct.SubCommand.html' title='clap::SubCommand'>SubCommand</a>&lt;'a&gt;</code></h3><div class='impl-items'><h4 id='method.new' class='method'><a class='stability Unmarked' title='No stability level'></a><code>fn <a href='#method.new' class='fnname'>new</a>&lt;'n, 'au, 'v, 'ab, 'u, 'ar&gt;(name: &amp;'n <a href='http://doc.rust-lang.org/nightly/std/primitive.str.html'>str</a>) -&gt; <a class='struct' href='../clap/struct.App.html' title='clap::App'>App</a>&lt;'au, 'v, 'ab, 'u, 'ar&gt;</code></h4>
<div class='docblock'><p>Creates a new instance of a subcommand requiring a name. Will be displayed
to the user when they print version or help and usage information.</p>

<h1 id="example" class='section-header'><a
                           href="#example">Example</a></h1><pre id='rust-example-rendered' class='rust '>
<span class='ident'>SubCommand</span>::<span class='ident'>new</span>(<span class='string'>&quot;config&quot;</span>)
</pre>
</div></div></section>
    <section id='search' class="content hidden"></section>

    <section class="footer"></section>

    <div id="help" class="hidden">
        <div class="shortcuts">
            <h1>Keyboard shortcuts</h1>
            <dl>
                <dt>?</dt>
                <dd>Show this help dialog</dd>
                <dt>S</dt>
                <dd>Focus the search field</dd>
                <dt>&larrb;</dt>
                <dd>Move up in search results</dd>
                <dt>&rarrb;</dt>
                <dd>Move down in search results</dd>
                <dt>&#9166;</dt>
                <dd>Go to active search result</dd>
            </dl>
        </div>
        <div class="infos">
            <h1>Search tricks</h1>
            <p>
                Prefix searches with a type followed by a colon (e.g.
                <code>fn:</code>) to restrict the search to a given type.
            </p>
            <p>
                Accepted types are: <code>fn</code>, <code>mod</code>,
                <code>struct</code>, <code>enum</code>,
                <code>trait</code>, <code>typedef</code> (or
                <code>tdef</code>).
            </p>
        </div>
    </div>

    

    <script>
        window.rootPath = "../";
        window.currentCrate = "clap";
        window.playgroundUrl = "";
    </script>
    <script src="../jquery.js"></script>
    <script src="../main.js"></script>
    
    <script async src="../search-index.js"></script>
</body>
</html>