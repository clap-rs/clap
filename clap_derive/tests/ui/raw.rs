// Copyright 2018 Guillaume Pinot (@TeXitoi) <texitoi@texitoi.eu>
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

use clap::Clap;

#[derive(Clap, Debug)]
struct Opt {
    #[clap(raw(case_insensitive = "true"))]
    s: String,
}

#[derive(Clap, Debug)]
struct Opt2 {
    #[clap(raw(requires_if = r#""one", "two""#))]
    s: String,
}
fn main() {
    let opt = Opt::parse();
    println!("{:?}", opt);
}
