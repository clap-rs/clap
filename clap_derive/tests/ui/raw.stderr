error: `#[clap(raw(...))` attributes are removed, they are replaced with raw methods

  = help: if you meant to call `clap::Arg::raw()` method you should use bool literal, like `raw(true)` or `raw(false)`
  = note: if you need to call `clap::Arg/App::case_insensitive` method you can do it like this: #[structopt(case_insensitive = true)]

  --> $DIR/raw.rs:13:12
   |
13 |     #[clap(raw(case_insensitive = "true"))]
   |            ^^^

error: `#[clap(raw(...))` attributes are removed, they are replaced with raw methods

  = help: if you meant to call `clap::Arg::raw()` method you should use bool literal, like `raw(true)` or `raw(false)`
  = note: if you need to call `clap::Arg/App::requires_if` method you can do it like this: #[structopt(requires_if("one", "two"))]

  --> $DIR/raw.rs:19:12
   |
19 |     #[clap(raw(requires_if = r#""one", "two""#))]
   |            ^^^

error[E0277]: the trait bound `Opt: std::convert::From<clap::parse::matches::arg_matches::ArgMatches>` is not satisfied
  --> $DIR/raw.rs:11:10
   |
11 | #[derive(Clap, Debug)]
   |          ^^^^ the trait `std::convert::From<clap::parse::matches::arg_matches::ArgMatches>` is not implemented for `Opt`

error[E0277]: the trait bound `Opt2: std::convert::From<clap::parse::matches::arg_matches::ArgMatches>` is not satisfied
  --> $DIR/raw.rs:17:10
   |
17 | #[derive(Clap, Debug)]
   |          ^^^^ the trait `std::convert::From<clap::parse::matches::arg_matches::ArgMatches>` is not implemented for `Opt2`
